### PR TITLE
Bug 1268423 - [TV][Home] Sort cards by alphabet order in the filtering mode

### PR DESCRIPTION
**Pull request by Fischer-L**
This patch would do in the filtering mode:
- Hide cards which has the same name and is launched to the same url
- Sort cards based on names

### CODE REVIEW

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=6fc1363bf934babcb6684eeef096d393b9f5584e

File: tv_apps/smart-home/js/filter_manager.js [hunk 1]
```diff
@@ -206,27 +206,54 @@
...
+      } else {
+        this._cardManager.getFilteredCardList(filter.name).then(list => {
+          var appearedList = [];
+          var uniqueList = [];
+          list.forEach(card => {
+            // Get url to which card is launched
+            var url = card.launchURL ? card.launchURL : card.manifest.origin;
+            // Get sorting key based on card name
+            var key = CardUtil.getSortKey(card);
+            // The same card which appeared before will not be picked up.
```
Comment (luke-chang) [inline tv_apps/smart-home/js/filter_manager.js]: nit: "The same card which **appears** before" OR "The same card **appearing** before"

File: tv_apps/smart-home/js/filter_manager.js [hunk 2]
```diff
@@ -206,27 +206,54 @@
...
+            // for checking here)
+            if (appearedList.indexOf(url + key) < 0) {
+              appearedList.push(url + key);
+              uniqueList.push({
+                card: card,
+                key: key
+              });
+            }
+          });
+          uniqueList.sort((a, b) => a.key.localeCompare(b.key));
```
Comment (luke-chang) [inline tv_apps/smart-home/js/filter_manager.js]: I would recommend you to remove duplicate cards **after** sorting for better performance. (although it doesn't really matter with that small scale, it's worth a try.)

Like this:

```js
var lastItem;
list.reduce(function(reducedArray, item) {
  if (item !== lastItem) {
    lastItem = item;
    reducedArray.push(item);
  }
  return reducedArray;
}, []);
```

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=5940c361ca4b1f6197feb622ed040542579f5a9b

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=71d0487b4c3cb7d91fdce8643fd4365ebc153b28

File: tv_apps/smart-home/js/filter_manager.js (hunk 1 above)
Comment (Fischer-L) [inline tv_apps/smart-home/js/filter_manager.js]: Updated to "The same card **appearing** before"

File: tv_apps/smart-home/js/filter_manager.js (hunk 2 above)
Comment (Fischer-L) [inline tv_apps/smart-home/js/filter_manager.js]: Thanks, updated to remove duplicate cards after sorting

File: tv_apps/smart-home/js/filter_manager.js [hunk 3]
```diff
@@ -206,27 +206,65 @@
...
+            var result = a.key.localeCompare(b.key);
+            if (result === 0) {
+              result = a.url.localeCompare(b.url);
+            }
+            return result;
+          });
+
+          var uniqueItems = [];
+          allItems.forEach(item => {
+            var last = uniqueItems[uniqueItems.length - 1];
```
Comment (luke-chang) [inline tv_apps/smart-home/js/filter_manager.js]: Let's try `last = item;` and put it at the last line.

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=62a7c01f74fb3cb880829bf2af33d2ecfcacb4ba

**Pull request merged**